types/connector.d.ts

```diff
@@ -3,28 +3,285 @@
+   * 
+   * @param modelName Target table/view name
+   * @param options Discovery options
+   * @param cb Callback function
+   */
+  discoverModelProperties?(modelName: string, options: DiscoveryScopeOptions, cb: Callback<DiscoveredModelProperties>): Promise<DiscoveredModelProperties>;
+  /**
+   * {@inheritDoc Connector.discoverModelProperties}
+   * @deprecated
+   */
+  discoverModelPropertiesSync?(modelName: string, options: DiscoveryScopeOptions): DiscoveredModelProperties;
+
+  /**
+   * Discover primary keys for a given owner/model name.
+   * 
+   * @param modelName Target model name
+   * @param options Discovery options
+   * @param cb Callback function
+   */
+  discoverPrimaryKeys?(modelName: string, options: IDPropertiesDiscoveryOptions, cb: Callback<DiscoveredPrimaryKeys>): Promise<DiscoveredPrimaryKeys>;
+
+  /**
+   * {@inheritDoc Connector.discoverPrimaryKeys}
+   * @deprecated
+   */
+  discoverPrimaryKeysSync?(modelName: string, options: IDPropertiesDiscoveryOptions): DiscoveredPrimaryKeys;
+
+  /**
+   * Discover foreign keys for a given owner/model name.
+   * 
+   * @param modelName Target model name
+   * @param options Discovery options
+   * @param cb Callback function
+   */
+  discoverForeignKeys?(modelName: string[], options: IDPropertiesDiscoveryOptions, cb: Callback<DiscoveredForeignKeys>): Promise<DiscoveredForeignKeys>;
+  /**
+   * {@inheritDoc Connector.discoverForeignKeys}
+   * @deprecated
+   */
+  discoverForeignKeysSync?(modelName: string[], options: IDPropertiesDiscoveryOptions): DiscoveredForeignKeys;
+
+  /**
+   * Retrieve a description of the foreign key columns that reference the given
+   * table's primary key columns (i.e. The foreign keys exported by a table),
+   * ordered by `fkOwner`, `fkTableName`, and `keySeq`.
+   * 
+   * @param modelName Target model name
+   * @param options Discovery options
+   * @param cb Callback function
+   */
+  discoverExportedForeignKeys?(modelName: string, options: IDPropertiesDiscoveryOptions, cb: Callback<DiscoveredForeignKeys>): Promise<DiscoveredForeignKeys>;
+
+  /**
+   * {@inheritDoc Connector/discoverExportedForeignKeys}
+   * @deprecated
+   */
+  discoverExportedForeignKeysSync?(modelName: string, options?: {owner?: string}): DiscoveredForeignKeys;
+
+  /**
+   * Discover schema from a given table name / view name.
+   * 
+   * @param tableName Target table name
+   * @param options Discovery options
+   * @param cb Callback function
+   */
+  discoverSchemas(tableName: string, options: SchemaDiscoveryOptions, cb: Callback<Schema>): Promise<Schema>;
+
+  /**
+   * Check whether or not migrations are required for the database schema to match
+   * the model definitions attached to the {@link DataSource}.
+   * 
+   * @param models Name of models to check. If not defined, all models are checked.
+   * @param cb 
+   */
+  isActual?(models?: string | string[], cb?: Callback<boolean>): void;
+  freezeDataSource?(): void;
+  freezeSchema?(): void;
```

Comment:
Not sure the difference between these functions. `DataSource.prototype.freeze` calls both functions if they are defined.

Reply:
It seems to have been introduced by this commit:

https://github.com/loopbackio/loopback-datasource-juggler/commit/9b169efc8c7277e69947239a7aff5bc64383e870#diff-7de38f47065ef67b4936fbd6396bb71a580fe9a5a9ff2c9723c4fe859d1e2a6eR977-R979

Reply:
Based on the commit just before the one mentioned above, they should be identical, and the use of both is for backwards-compatibility with JugglingDB:

https://github.com/loopbackio/loopback-datasource-juggler/commit/6af4b1ba2f5894c32d960258f43b38f2b69a37e1